app/src/main/java/com/example/weatherapp/models/MainActivity.kt

```diff
@@ -2,24 +2,34 @@ package com.example.weatherapp.models
 
 import androidx.appcompat.app.AppCompatActivity
 import android.os.Bundle
+import androidx.activity.viewModels
+import androidx.lifecycle.lifecycleScope
 import androidx.recyclerview.widget.LinearLayoutManager
 import androidx.recyclerview.widget.RecyclerView
 import com.example.weatherapp.R
+import kotlinx.coroutines.launch
 
 class MainActivity : AppCompatActivity() {
 
     private lateinit var weatherAdapter: WeatherListAdapter
+    private val weatherMainMenuViewModel: WeatherMainMenuViewModel by viewModels()
+
+    private val defaultList = listOf<Location>()
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
         initRecyclerView()
-        addDataSet()
+        lifecycleScope.launch{
+            setDataSet()
+            weatherMainMenuViewModel.uiState.collect { locationData ->
+                weatherAdapter.submitList(locationData.locations?: defaultList)
+            }
+        }
     }
 
-    private fun addDataSet(){
-        val data = WeatherDataSource.createWeatherDataSet()
-        weatherAdapter.submitList(data)
+    private suspend fun setDataSet(){
+        weatherMainMenuViewModel.setListOfCountries()
```

Comment:
Similar comment as the one from the DetailsActivity where the view is the one acting on the ViewModel, and with a suspend function.

This one seems to be simpler to solve. Maybe just do the action when the ViewModel is constructed?

Reply:
Thanks for the idea Allan, I will try this right now 
Gotta make the view do less work 🕺